### DISCUSSION TITLE
Why does prisma select each individual column instead of a * when no select clause is included?

### DISCUSSION THREAD

**Discussion by MorenoMdz** [prisma/prisma#23548] [Q&A] [answered]
### Question

Hey there, I want to clarify one situation. 

We had a problem running a migration where one column was being used, but during the time we deployed the migration and the code changes were deployed some queries failed, and these were all "find/find many" queries that didn't have a list of selected fields. Checking the queries that ran from the DB side it seems Prisma is doing a select in all the columns in the schema instead of doing a `SELECT *`.

If we got it correct, the fact is that we are going to have to run a two steps deployment in these cases, one to update the schema so the type doesn't have that column anymore, and another deploy/release to actually migrate the DB to DROP the column, which is kinda annoying considering that we need to run those back to back before any other migrations are generated.

A couple things I need to clarify:
* Does Prisma actually run individual select for all columns when no select clause is provided? If so is there a reason for that, like performance or type safety?
* Is there a recommended way to run these migrations that don't force us to have two releases back to back?

**Comment by ludralph** [comment, marked as answer]
Hi @MorenoMdz 👋 

Thank you for raising this question.

> Does Prisma actually run individual select for all columns when no select clause is provided? If so is there a reason for that, like performance or type safety?

Yes, Prisma Client does run individual select for all columns when no select clause is provided. By default, when a query returns records, the result includes all scalar fields defined in the Prisma schema (including enums) and none of the relations. This is done for type safety reasons. Prisma Client generates the return type for its queries on the fly. This means it's impossible for you to accidentally access a property on a model that has not been retrieved in a query.

> Is there a recommended way to run these migrations that don't force us to have two releases back to back?

Yes, there is a strategy you can adopt to avoid having to do two releases back to back when running migrations:

Expand and Contract Pattern: This pattern involves making all migrations backwards compatible, at least for a certain period of time. For example, if you need to rename a column, you could add a new column with the new name, copy the data from the old column to the new one, and then keep the old column for a while. This way, services using the old client version will still function correctly. Then, once all services are updated to use the new client version, you can remove the old column.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by MorenoMdz** [reply]
Thanks, wanted to confirm these assumptions. On the migration part, does it make sense to use the `@ignore` flag on the field in order to achieve the same result of the expand and contract pattern?

Keep in mind the edge case I mentioned the column that was removed from the schema was not in use anywhere, we had issues because the schema had the field it, and since Prisma runs a SELECT on every column in the model, vs a *, it tried to access a column that wasn't in the DB anymore, as we ran Migration -> Deploy, and there was a gap of few minutes in between.

**Reply by ludralph** [reply]
Yes it does. See this [comment](https://github.com/prisma/prisma/issues/4727#issuecomment-853728721)
